### PR TITLE
Tests for watch tasks

### PR DESCRIPTION
**Pull request by gauntface**
@surma thoughts on this? Still worth doing?

### CODE REVIEW

File: test/task-checker.js [hunk 1]
```diff
@@ -86,25 +84,17 @@ let describeTestsForTask = function(taskName, task) {
...
-    tasksToTest.push({
-      taskName: taskFilename,
-      task: require(path.join(TASKS_DIRECTORY, taskFilename))
-    });
-  });
+  const del = require('del');
 
   // Clean up before each test
-  beforeEach(done => rimraf(TEST_OUTPUT_PATH, done));
+  beforeEach(done => del(TEST_OUTPUT_PATH + '/**').then(() => done(), done));
```
Comment (surma) [inline test/task-checker.js]: Why is there `.then(() => done(), done)` instead of `.then(done, done)`?

File: test/helpers/task-helper.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
+const fs = require('fs');
```
Comment (surma) [inline test/helpers/task-helper.js]: Why `require` instead of `import`?

File: test/watch-test.js [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+const TEST_OUTPUT_PATH = path.join('test', 'output');
+const TEST_OUTPUT_SRC = path.join(TEST_OUTPUT_PATH, 'src');
+const TEST_OUTPUT_DEST = path.join(TEST_OUTPUT_PATH, 'build');
+
+let watcherTask;
+
+// Use rimraf over del because it seems to work more reliably on Windows.
+// Probably due to it's retries.
+const deleteFiles = path => {
+  return new Promise((resolve, reject) => {
```
Comment (surma) [inline test/watch-test.js]: No need for `return`

File: test/watch-test.js [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+    try {
+      fs.lstatSync(path.join(TEST_OUTPUT_SRC, folderName));
+    } catch (err) {
+      // Path doesn't exist, we can ignore it
+      return;
+    }
+
+    const expectedOutputFileBuffer = fs.readFileSync(path.join(TEST_OUTPUT_SRC, folderName, 'output.json'));
+    const expectedOutput = JSON.parse(expectedOutputFileBuffer.toString());
+    expectedOutput.forEach(file => {
```
Comment (surma) [inline test/watch-test.js]: So we don’t care about extra files in the output?

File: test/watch-test.js [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+      }
+
+      if (pathstats.size <= 0) {
+        throw new Error(`Output file has no contents: ${fullpath}`);
+      }
+    });
+  });
+};
+
+const performStep = (step, watcher) => {
```
Comment (surma) [inline test/watch-test.js]: Please add a comment this function. It’s kinda hard to grok.

File: test/watch-test.js [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+      } else {
+        setTimeout(timeoutHandler, 1000);
+      }
+    };
+
+    timeoutHandler();
+  });
+};
+
+const stepOverEachStep = (steps, watcher) => {
```
Comment (surma) [inline test/watch-test.js]: Same.

File: test/task-checker.js (hunk 1 above)
Comment (surma) [inline test/task-checker.js]: nvm. Got it. `done(something)` is for errors, isn’t it?

File: test/watch-test.js [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+      // This is a long time to account for slow babel builds on Windows
+      this.timeout(60000);
+
+      const steps = [
+        () => copyFiles(VALID_TEST_FILES, TEST_OUTPUT_SRC)
+      ];
+
+      performTest(taskName, task, steps)
+      .then(() => done())
+      .catch(err => done(err));
```
Comment (surma) [inline test/watch-test.js]: `.then(() => done(), done)`?

File: test/watch-test.js [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+      resolve();
+    });
+  });
+};
+
+const performTest = (taskName, task, steps) => {
+  // Start the tasks watching
+  watcherTask = task.watch();
+  if (!watcherTask) {
+    return Promise.reject(new Error(`Nothing returned from the tasks watch() method. Is the result of gulp.watch returned in ${taskName}`));
```
Comment (surma) [inline test/watch-test.js]: add `?` to msg

File: test/watch-test.js [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+      ];
+
+      performTest(taskName, task, steps)
+      .then(() => done())
+      .catch(err => done(err));
+    });
+
+    it('should watch for new files being added and changed', function(done) {
+      // This is a long time to account for slow babel builds on Windows
+      this.timeout(60000);
```
Comment (surma) [inline test/watch-test.js]: Can we remove this as we are skipping the tests on windows anyways?

File: test/task-checker.js (hunk 1 above)
Comment (gauntface) [inline test/task-checker.js]: Yup :(

File: test/helpers/task-helper.js (hunk 2 above)
Comment (gauntface) [inline test/helpers/task-helper.js]: import isn't in node is it?

File: test/watch-test.js (hunk 3 above)
Comment (gauntface) [inline test/watch-test.js]: There is otherwise there will be no promise to bing to.

File: test/watch-test.js (hunk 4 above)
Comment (gauntface) [inline test/watch-test.js]: Not unless we can't to care about deleting files as we go along (i.e. a watch kicks off a clean and then a build).

That opens issues with sass around cleaning CSS files up.

File: test/helpers/task-helper.js (hunk 2 above)
Comment (surma) [inline test/helpers/task-helper.js]: lols. right

File: test/watch-test.js (hunk 3 above)
Comment (surma) [inline test/watch-test.js]: What I meant was: Use `path => new Promise(...)` instead of `path => { return new Promise(...); }`

**Comment by gauntface** [comment]
@surma fixed the tests - sorry for the delay with that

**Comment by surma** [comment]
@gauntface LGTM. I think we should include this, so merge at will.

**Pull request merged**